### PR TITLE
Issue #45: Add Camelot Nitro Pool Interface to SDK

### LINKED ISSUE #45: Add Camelot Nitro Pool Interface to SDK
Part 1 of TVL project https://github.com/open-dollar/od-app/issues/234

- Need to connect to Camelot nitro pools via an interface

Potentially helpful resources: 
https://github.com/Zapper-fi/studio/blob/bd5d0f3d851b18c5dcd9c129db24312621042594/src/apps/camelot/arbitrum/camelot.nitro.contract-position-fetcher.ts#L18
https://github.com/lyra-finance/interface/blob/59661df7111ccb838398a47152be13144f73f354/app/src/contracts/typechain/CamelotNitroPool.ts#L51

### LINKED ISSUE #53: Enable multiple pools
Depends on #45
 
Now that we can load data from a single pool, we need to add the ability to pull for multiple pools.

- Rename `fetchNitroPoolODGwstETH` to `fetchNitroPool` which is more generic
- The arguments should accept the token symbol (eg. WSTETH), instead of the address of the pool
- Validate that we are fetching all data needed (see both app screenshots below) 

image

image

### LINKED ISSUE #54: Fetch market price using web3
Depends on #53 

- Swap out the use of `oracleRelayer.marketPrice()` to get the ODG price, for the actual camelotRelayer we will deploy (depends on https://github.com/open-dollar/od-relayer/issues/33)
- Get the correct price for WSTETH using the chainlink oracle

### LINKED ISSUE #55: Fetch market price using web2
Depends on #53

Faster alternative to #54 

Use Alchemy or another SDK to fetch price data for WSTETH, RETH, and ODG

### LINKED ISSUE #56: Update mainnet collateral addresses
### WSTETH
0x5979d7b546e38e414f7e9822514be443a4800529 
https://app.aave.com/reserve-overview/?underlyingAsset=0x5979d7b546e38e414f7e9822514be443a4800529&marketName=proto_arbitrum_v3
Oracle CLSynchronicityPriceAdapterPegToBase: https://arbiscan.io/address/0x945fD405773973d286De54E44649cc0d9e264F78

### ARB 
0x912ce59144191c1204e64559fe8253a0e49e6548
https://app.aave.com/reserve-overview/?underlyingAsset=0x912ce59144191c1204e64559fe8253a0e49e6548&marketName=proto_arbitrum_v3
Oracle EACAggregatorProxy: https://arbiscan.io/address/0xb2A824043730FE05F3DA2efaFa1CBbe83fa548D6

### RETH
0xEC70Dcb4A1EFa46b8F2D97C310C9c4790ba5ffA8 
Not on camelot
https://app.aave.com/reserve-overview/?underlyingAsset=0xec70dcb4a1efa46b8f2d97c310c9c4790ba5ffa8&marketName=proto_arbitrum_v3
Oracle CLSynchronicityPriceAdapterPegToBase: https://arbiscan.io/address/0x04c28D6fE897859153eA753f986cc249Bf064f71

### CBETH: 
Not on camelot

### WETH
0x82af49447d8a07e3bd95bd0d56f35241523fbab1
https://app.aave.com/reserve-overview/?underlyingAsset=0x82af49447d8a07e3bd95bd0d56f35241523fbab1&marketName=proto_arbitrum_v3
Oracle EACAggregatorProxy: https://arbiscan.io/address/0x639Fe6ab55C921f74e7fac1ee960C0B6293ba612

### PR DESCRIPTION
**Pull request by jahabeebs**
Closes #45 
Closes #56 
Closes #53 
Closes #54 
Closes #55 

## Description
- I added the function fetchNitroPoolODGwstETH (...maybe I should snake case this) which fetches pool data from an ODG-WSTETH nitro pool and returns a lot of relevant stats
- I added several new contracts to our contract API and GEB object (note that I added the ABIs into our local folder)--this will allow us to generate typechained definitions for these contracts and interact with them in the app easily. These contracts include:
1) camelotNitroPool, the camelot nitro pool ABI which I've taken from Lyra Finance (https://github.com/lyra-finance/interface/commit/e60e7bcf9d43ebafa8ee1b9d6ba8820e7572d00e#diff-01893f21d6faed4a7146ec796292868d585f39615ccd5787961a9e5a4e5f65da), I'm assuming it's up to date for now until we deploy our own
2) camelotODGPool, this is of type CamelotPool which is the ABI of our *deployed ODG pool*. We need this so that we can access market price info for ODG. However, there's a problem with this contract which I'll note in a section below
3) mutlicall, this is a contract that lets us create batch calls on any network--we already had the addresses for this multicall contract in our addresses.ts file, however it wasn't available on the GEB object or our contract APIs so I used the official multicall site to get the ABI and this will let us do batch calls in the app

## Problems/questions

We're missing three key pieces of info for testing this code
1) No deployed Arbitrum Sepolia Camelot pools for ODG-wstETH or ODG-wETH. I've added a TODO in addresses.ts for adding these two contracts 
2) Need chainlinkRelayer for wstETH—in fetchPoolData we're using chainlinkRelayer to get the wETH market price. However, we do not have a chainlinkRelayer for Arbitrum Sepolia, just Arbitrum Goerli, which which is probably why failures occur with fetching pool data in our app
3) The deployed CamelotPool contract we're using for ODG has a lot of functions but not totalSupply. We need a Camelot pool contract with totalSupply as part of our multicall logic. I don't know why totalSupply is missing from our contract -- maybe we're using a new version? The CamelotPool ABI that Lyra is using just has totalSupply (https://github.com/lyra-finance/interface/blob/e60e7bcf9d43ebafa8ee1b9d6ba8820e7572d00e/app/src/contracts/abis/CamelotPool.json). Is there an alternative way to calculate this? I've added a TODO in fetchNitroPool.ts to remind us that the ABI needs to be addressed

Also, we'll later need to create a function for fetching from the RETH pool, but we should get this one working first

### CODE REVIEW

File: src/contracts/addreses.ts [hunk 1]
```diff
@@ -78,6 +80,9 @@ const addresses: Record<GebDeployment, ContractList> = {
         JOB_ACCOUNTING: '0x5178C1A616CaF67184104Cc6555c0310bbB7F6d5',
         JOB_LIQUIDATION: '0xD1D1a18a459d9Be1a7C237578DF1aB3CFf051282',
         JOB_ORACLES: '0x67cAD916b2572708E352eDE3A16E54F8d353ebDB',
+        // TODO: Add nitro and ODG pool addresses
+        CAMELOT_NITRO_POOL: '',
+        CAMELOT_ODG_POOL: ''
     },
     'arbitrum-goerli': {
```
Comment (pi0neerpat) [inline src/contracts/addreses.ts]: Are you on the latest branch? I thought I removed `arbitrum-goerli`

File: src/contracts/addreses.ts [hunk 2]
```diff
@@ -37,6 +37,8 @@ export type ContractKey =
     | 'JOB_ACCOUNTING'
     | 'JOB_LIQUIDATION'
     | 'JOB_ORACLES'
+    | 'CAMELOT_NITRO_POOL'
```
Comment (pi0neerpat) [inline src/contracts/addreses.ts]: Will need to differentiate Nitro pools by collateral types: `RETH`, `WSTETH`

File: src/camelot/utils.ts
Comment (pi0neerpat) [inline src/camelot/utils.ts:1]: Maybe move to `src/`, since this may be used for other features

**Review by pi0neerpat — CHANGES_REQUESTED** [review]

File: src/contracts/addreses.ts (hunk 1 above)
Comment (jahabeebs) [reply src/contracts/addreses.ts]: yes, goerli is still in dev

File: src/contracts/addreses.ts (hunk 2 above)
Comment (jahabeebs) [reply src/contracts/addreses.ts]: hmm good thought, I'll add that in

**Comment by jahabeebs** [comment]
Pushed several updates:
- Gets rid of totalSupply logic
- New (rough) APY calculation: ```(rewardsPerSecond * SECONDS_IN_YEAR * odgMarketPriceFloat) / tvl```
- Updates Nitro Pool ABI with more functions, now we get this pool-level data:
```
   settings: {
        startTime: BigNumber
        endTime: BigNumber
        harvestStartTime: BigNumber
        depositEndTime: BigNumber
        lockDurationReq: BigNumber
        lockEndReq: BigNumber
        depositAmountReq: BigNumber
        whitelist: boolean
        description: string
    }
```
- Returns the entire userInfo object to get important deposit info:
```
    userInfo: {
        totalDepositAmount: BigNumber,
        rewardDebtToken1: BigNumber,
        rewardDebtToken2: BigNumber,
        pendingRewardsToken1: BigNumber,
        pendingRewardsToken2: BigNumber
    } | null,
```

**Comment by jahabeebs** [comment]
Pushed several updates:
- Removed camelotODGPool (we will have wsteth and reth pools, this is not needed)
- Renamed MulticallRequest to CamelotMulticallRequest
- balanceOf logic changed in fetchNitroPool logic to pass in correct camelotwstETHNitroPool address to the test ODG and wstETH contracts
- Addresses updated to include the correct test contracts (like the wstETH pool, test wstETH, and test ODG addresses)

**Review by pi0neerpat — CHANGES_REQUESTED** [review]
I pushed some changes to get this to display details about the pool in the app. See https://github.com/open-dollar/od-app/pull/252 

TODO

- Add back data for `lpTokenBalance` and `userInfo`, which I temporarily removed while debugging
- The naming convention for the SDK is to make token symbols always uppercase. Eg fetchNitroPool wstETH_market_price should be changed to WSTETH_market_price

Other comments that I already addressed:

- Use 0x0 addresses instead of a real address from the wrong network. This prevents confusion in the future.

**Comment by jahabeebs** [comment]
@pi0neerpat 
- Added back data for lpTokenBalance and userInfo
- Renamed wstETH to WSTETH throughout my code so that it adheres to standards

Here's what the stringified data looks like now in the app

Screenshot 2024-02-05 at 2 25 22 PM

**Review by pi0neerpat — COMMENTED** [review]
What data is left to gather based on the designs?

image

**Comment by jahabeebs** [comment]
> What data is left to gather based on the designs?
> 
> image

Nothing, this has 100% of the data needed

We just need to replace the price data with real values

**Comment by jahabeebs** [comment]
I pushed changes to this PR that refactor the nitro pool method to fetch data for our RETH or WSTETH test pools. so in the app for the RETH-ODG pool you would call it like 

```
const data = await fetchNitroPool(geb, 'RETH','0xaddresshere')
```

A strange thing that happened was when testing today I was getting errors pulling from my wstETH pool until I re-imported the token to camelot's UI, might need to do some more testing to see if pulling from our pools long-term relies on Camelot allow listing a token

This PR now closes #53 too

## Screenshots

https://github.com/open-dollar/od-sdk/assets/47253537/c2c5ac68-2389-497e-aa63-4c7e779eec18

**Comment by jahabeebs** [comment]
Pushed an update that includes the following:

1) no hardcoded prices--RETH, ODG, and WSTETH prices are being fetched from a USD denominated oracle contract that I added to ```addresses.ts```. The only issue here is that the validity of RETH and WSTETH is coming back as false so I had to temporarily disable the validity check of these prices. This should be ok for now since it's read-only data
2) ```availableCollateralsForDeposit``` function returns an array of available collaterals for deposit
3) added pendingRewards object to response which includes pending1 and pending2 values...there are two values within the object because through the Camelot UI we can add incentives twice. in the example pools I've only added incentives once so only pending1 will be populated
4) Added type AvailableDepositTypes to prevent type casting in the app

This now closes #54 and #55--we didn't need to use the coingecko fetch since now we have USD denominated oracle contracts

**Comment by jahabeebs** [comment]
Closing as won't do -- part of TVL project

**Pull request closed without merging**